# Finding differentially expressed genes and isoforms {style="gray"}

## Differentially expressed genes by cluster identity 

first we will look at marker genes for each cluster. This will help us identify which genes are DE in each cluster and indicate which cell type each cluster belongs to. We will also look at DE isoforms using the same methodology

```{r perform_marker_gene_analysis, echo=TRUE, message=FALSE, warning=FALSE, echo=FALSE}
#Find markers for all clusters using the "RNA" and "iso" assay
#One

all_markers_gene <- FindAllMarkers(seu_obj, assay = "RNA", do.print = TRUE,
                                       logfc.threshold = 0.5, min.pct = 0.20, only.pos = TRUE) %>% filter(p_val_adj < 0.05)

all_markers_iso <- FindAllMarkers(seu_obj, assay = "iso", do.print = TRUE,
                                      logfc.threshold = 0.5, min.pct = 0.20, only.pos = TRUE) %>% filter(p_val_adj < 0.05)

#save the list of DE genes 
write.csv(all_markers_one_gene, "./DE_genes_and_isoforms/all_markers_one_gene.csv")
write.csv(all_markers_one_iso, "./DE_genes_and_isoforms/all_markers_one_iso.csv")

```

# Footnotes and citations

## Footnotes

Footnotes are put inside the square brackets after a caret `^[]`. Like this one [^finding_differentially_expressed_genes_and_isoforms-1].

[^finding_differentially_expressed_genes_and_isoforms-1]: This is a footnote.

## Citations

Reference items in your bibliography file(s) using `@key`.

For example, we are using the **bookdown** package [@R-bookdown] (check out the last code chunk in index.Rmd to see how this citation key was added) in this sample book, which was built on top of R Markdown and **knitr** [@xie2015] (this citation was added manually in an external file book.bib). Note that the `.bib` files need to be listed in the index.Rmd with the YAML `bibliography` key.

The RStudio Visual Markdown Editor can also make it easier to insert citations: <https://rstudio.github.io/visual-markdown-editing/#/citations>
